services/src/main/java/org/keycloak/services/HttpRequestImpl.java

```diff
@@ -61,8 +65,19 @@ public MultivaluedMap<String, String> getDecodedFormParameters() {
         }
         MediaType mediaType = getHttpHeaders().getMediaType();
         if (mediaType == null || !mediaType.isCompatible(MediaType.valueOf("application/x-www-form-urlencoded"))) {
-            return new MultivaluedHashMap<>();
+            throw new BadRequestException("MediaType is not compatible");
+        } 
+        
+        String charsetName = mediaType.getParameters().get(MediaType.CHARSET_PARAMETER);
+        
+        if(charsetName != null) {
+            try {
+                Charset.isSupported(charsetName);
+            } catch (Exception e) {
+                throw new BadRequestException("Charset " + charsetName +" is not supported");
+            }
         }
+        
         return delegate.getDecodedFormParameters();
```

Comment:
I think the charset is going to be OK 99.99% of the times. Reasteasy is also checking the charset so I was thinking in saving that time, delete the check completely and just use the general catch later. The message will be something like: `Incorrect form parameters: java.io.UnsupportedEncodingException: 'utf-8'`.  It could be better but that's not our fault. WDYT?

Reply:
I don't know why, but without the following, the RestEasy is not validating the `'utf-8'`:

```java
            if(charsetName != null) {
                Charset.isSupported(charsetName);
            }
```

and we get the message: `"Incorrect form parameters: 'utf-8'"`. However, if we test with `utf-88`, than the message is: `Incorrect form parameters: java.io.UnsupportedEncodingException: utf-88`. So, I need to keep that validation.

Reply:
Maybe it's because you are re-catching the exception. What happens if you just do the following?

```java
        MediaType mediaType = getHttpHeaders().getMediaType();
        if (mediaType == null || !mediaType.isCompatible(MediaType.valueOf("application/x-www-form-urlencoded"))) {
            throw new BadRequestException("MediaType " + mediaType + " is not application/x-www-form-urlencoded compatible");
        }
        try {
            return delegate.getDecodedFormParameters();
        } catch (Exception e) {
            throw new BadRequestException("Incorrect form parameters: " + e.getMessage(), e);
        }
```

I's say that the message should be: `Incorrect form parameters: java.io.UnsupportedEncodingException: 'utf-8'`.

Reply:
Hummmm. I gesso that I have tried this also. Anyway, I will try again. Thanks